x-pack/plugins/observability_solution/inventory/common/utils/join_by_key.ts

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { UnionToIntersection, ValuesType } from 'utility-types';
+import { merge, castArray } from 'lodash';
+import stableStringify from 'json-stable-stringify';
+
+export type JoinedReturnType<
+  T extends Record<string, any>,
+  U extends UnionToIntersection<T>
+> = Array<
+  Partial<U> & {
+    [k in keyof T]: T[k];
+  }
+>;
+
+type ArrayOrSingle<T> = T | T[];
+
+export function joinByKey<
+  T extends Record<string, any>,
+  U extends UnionToIntersection<T>,
+  V extends ArrayOrSingle<keyof T & keyof U>
+>(items: T[], key: V): JoinedReturnType<T, U>;
+
+export function joinByKey<
+  T extends Record<string, any>,
+  U extends UnionToIntersection<T>,
+  V extends ArrayOrSingle<keyof T & keyof U>,
+  W extends JoinedReturnType<T, U>,
+  X extends (a: T, b: T) => ValuesType<W>
+>(items: T[], key: V, mergeFn: X): W;
+
+// TODO util already exists in apm and other plugins. Move it to oblt shared
+export function joinByKey(
```

Comment:
Can we move this to `observability_utils` instead? This is exactly like the one used on APM.

Reply:
https://github.com/elastic/kibana/pull/195250/commits/6800888ce55f5c873bd703d274a8efb6d1d3c21b